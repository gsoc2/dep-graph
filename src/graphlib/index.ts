export { Graph } from './graph';
import { isAcyclic } from './alg/is-acyclic';
import { postorder } from './alg/postorder';

export const alg = {
  isAcyclic,
  postorder,
};
