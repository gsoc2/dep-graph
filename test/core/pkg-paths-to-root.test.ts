import { createFromJSON, DepGraphData } from '../../src';
import { loadFixture } from '../helpers';

describe('pkgPathsToRoot', () => {
  it('calculates paths from a package to the root', () => {
    const graphJson: DepGraphData = loadFixture('simple-graph.json');

    const depGraph = createFromJSON(graphJson);

    const paths = depGraph.pkgPathsToRoot({
      name: 'e',
      version: '5.0.0',
    });

    expect(paths).toEqual([
      [
        {
          name: 'e',
          version: '5.0.0',
        },
        {
          name: 'd',
          version: '0.0.1',
        },
        {
          name: 'c',
          version: '1.0.0',
        },
        {
          name: 'a',
          version: '1.0.0',
        },
        {
          name: 'root',
          version: '0.0.0',
        },
      ],
      [
        {
          name: 'e',
          version: '5.0.0',
        },
        {
          name: 'd',
          version: '0.0.2',
        },
        {
          name: 'c',
          version: '1.0.0',
        },
        {
          name: 'b',
          version: '1.0.0',
        },
        {
          name: 'root',
          version: '0.0.0',
        },
      ],
    ]);
  });

  it('can limit the number of paths returned', () => {
    const graphJson: DepGraphData = loadFixture('simple-graph.json');
    const depGraph = createFromJSON(graphJson);
    const limit = 1;

    const pathsWithoutLimit = depGraph.pkgPathsToRoot({
      name: 'e',
      version: '5.0.0',
    });

    const pathsWithlimit = depGraph.pkgPathsToRoot(
      {
        name: 'e',
        version: '5.0.0',
      },
      { limit },
    );

    expect(pathsWithoutLimit.length).toBeGreaterThan(limit);
    expect(pathsWithlimit).toHaveLength(limit);
  });

  describe('cycles', () => {
    const depGraphData = loadFixture('cyclic-complex-dep-graph.json');
    const depGraph = createFromJSON(depGraphData);
    it('returns expected paths for all packages', () => {
      depGraph.getPkgs().forEach((pkg) => {
        const pkgPathsToRoot = depGraph.pkgPathsToRoot(pkg);
        expect(pkgPathsToRoot).toMatchSnapshot(`${pkg.name}@${pkg.version}`);
      });
    });
  });
});
