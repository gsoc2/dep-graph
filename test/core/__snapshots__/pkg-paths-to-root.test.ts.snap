// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`pkgPathsToRoot cycles returns expected paths for all packages: a@1 1`] = `
Array [
  Array [
    Object {
      "name": "a",
      "version": "1",
    },
  ],
]
`;

exports[`pkgPathsToRoot cycles returns expected paths for all packages: b@2 1`] = `
Array [
  Array [
    Object {
      "name": "b",
      "version": "2",
    },
    Object {
      "name": "a",
      "version": "1",
    },
  ],
]
`;

exports[`pkgPathsToRoot cycles returns expected paths for all packages: c@3 1`] = `
Array [
  Array [
    Object {
      "name": "c",
      "version": "3",
    },
    Object {
      "name": "a",
      "version": "1",
    },
  ],
]
`;

exports[`pkgPathsToRoot cycles returns expected paths for all packages: d@4 1`] = `
Array [
  Array [
    Object {
      "name": "d",
      "version": "4",
    },
    Object {
      "name": "a",
      "version": "1",
    },
  ],
]
`;

exports[`pkgPathsToRoot cycles returns expected paths for all packages: e@5 1`] = `
Array [
  Array [
    Object {
      "name": "e",
      "version": "5",
    },
    Object {
      "name": "b",
      "version": "2",
    },
    Object {
      "name": "a",
      "version": "1",
    },
  ],
  Array [
    Object {
      "name": "e",
      "version": "5",
    },
    Object {
      "name": "c",
      "version": "3",
    },
    Object {
      "name": "a",
      "version": "1",
    },
  ],
  Array [
    Object {
      "name": "e",
      "version": "5",
    },
    Object {
      "name": "g",
      "version": "7",
    },
    Object {
      "name": "f",
      "version": "6",
    },
    Object {
      "name": "d",
      "version": "4",
    },
    Object {
      "name": "a",
      "version": "1",
    },
  ],
]
`;

exports[`pkgPathsToRoot cycles returns expected paths for all packages: f@6 1`] = `
Array [
  Array [
    Object {
      "name": "f",
      "version": "6",
    },
    Object {
      "name": "d",
      "version": "4",
    },
    Object {
      "name": "a",
      "version": "1",
    },
  ],
  Array [
    Object {
      "name": "f",
      "version": "6",
    },
    Object {
      "name": "e",
      "version": "5",
    },
    Object {
      "name": "b",
      "version": "2",
    },
    Object {
      "name": "a",
      "version": "1",
    },
  ],
  Array [
    Object {
      "name": "f",
      "version": "6",
    },
    Object {
      "name": "e",
      "version": "5",
    },
    Object {
      "name": "c",
      "version": "3",
    },
    Object {
      "name": "a",
      "version": "1",
    },
  ],
]
`;

exports[`pkgPathsToRoot cycles returns expected paths for all packages: g@7 1`] = `
Array [
  Array [
    Object {
      "name": "g",
      "version": "7",
    },
    Object {
      "name": "f",
      "version": "6",
    },
    Object {
      "name": "d",
      "version": "4",
    },
    Object {
      "name": "a",
      "version": "1",
    },
  ],
  Array [
    Object {
      "name": "g",
      "version": "7",
    },
    Object {
      "name": "f",
      "version": "6",
    },
    Object {
      "name": "e",
      "version": "5",
    },
    Object {
      "name": "b",
      "version": "2",
    },
    Object {
      "name": "a",
      "version": "1",
    },
  ],
  Array [
    Object {
      "name": "g",
      "version": "7",
    },
    Object {
      "name": "f",
      "version": "6",
    },
    Object {
      "name": "e",
      "version": "5",
    },
    Object {
      "name": "c",
      "version": "3",
    },
    Object {
      "name": "a",
      "version": "1",
    },
  ],
]
`;
